fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -115,6 +129,11 @@ def forward(self, *inputs: Any, **kwargs: Any) -> Any:
         # Reset all the grad reduce and bucket state flags
         self._grad_to_be_reduced = [True] * len(self._grad_to_be_reduced)
 
+        # Register a layer in the autograd which will fire the gatekeeper during the backward pass
+        sentinel = torch.zeros((1))  # This makes sure that this function is properly registered
+        sentinel.requires_grad = True
+        inputs = self._sync_layer.apply(self, sentinel, *inputs)
```

Comment:
This looks ok to me.
A couple of questions though:
- Is this module supposed to allow re-entrant autograd (calling backward within the backward, which `utils.checkpoint` is doing)?
- Is there a reason not to use a similar logic as DDP to detect un-used parameters? I am not saying their strategy is better, I am just wondering if there are key differences between the two?

Reply:
Thanks for the feedback ! Really a draft, testing out how/if that works, sorry for the unused code lost in the middle.
- re-entrant: I did not think of that to be honest, I don't think that it would work well by default, but I would need to understand more abut the expected behaviour in that case to think about remediations. When a second backward is called in the middle of a first one,  is that to get a second derivative or to recompute gradients which have been dropped from the first pass (given the checkpoint utilisation, guess is second hypothesis)
- re: DDP logic to find unused parameters, it's something we discussed with @mrshenli, I think that the added cost for every FW is a shame though, so I'm trying to make sure that there's no other solution before going down that route

Reply:
re-entrant is to compute gradients for an independent graph while your compute this backward. That is how checkpoint uses it indeed.

What would be the difference between this in practice vs the one in DDP?

Reply:
(not sure whether this branch will live on, but the reentrant question is interesting) This moves the gradient to specific ranks instead of all-reducing (and clears the gradients if possible after the reducing is done), that's the only real difference with DDP unfortunately (longer term plan is to ditch any somethingDDP code here and fold back to pytorch DDP)